mpc/share-conversion-core/src/a2m.rs

```diff
@@ -0,0 +1,91 @@
+//! This module implements the A2M algorithm.
+
+use super::{MulShare, OTEnvelope, ShareConvert};
+use crate::{fields::Field, ShareConversionCoreError};
+use rand::{CryptoRng, Rng};
+
+/// An additive share of `A = x + y`
+#[derive(Clone, Copy, Debug)]
+pub struct AddShare<T>(T);
+
+impl<T: Field> AddShare<T> {
+    /// Turn into a multiplicative share and get values for OT
+    ///
+    /// This function returns
+    ///   * `MulShare` - The sender's multiplicative share
+    ///   * `OTEnvelope` - Used for oblivious transfer
+    pub fn convert_to_multiplicative<R: Rng + CryptoRng>(
+        &self,
+        rng: &mut R,
+    ) -> Result<(MulShare<T>, OTEnvelope<T>), ShareConversionCoreError> {
+        // We need to exclude 0 here, because it does not have an inverse
+        // which is needed later
+        let random: T = loop {
+            let r = T::rand(rng);
+            if r != T::zero() {
+                break r;
+            }
+        };
+        let mut masks: Vec<T> = vec![T::zero(); T::BIT_SIZE as usize];
+        masks.iter_mut().for_each(|x| *x = T::rand(rng));
+
+        // set the last mask such that the sum of all 128 masks equals 0
+        masks[T::BIT_SIZE as usize - 1] = -masks
+            .iter()
+            .take(T::BIT_SIZE as usize - 1)
+            .fold(T::zero(), |acc, i| acc + *i);
+
+        // the inverse of the random share will be the multiplicative share for the sender
+        let mul_share = MulShare::new(random.inverse());
+
+        // decompose the share into a sum of components, e.g. if the share is 10110, we decompose it into
+        // 0 + 10 + 100 + 0000 + 10000
+        let components: Vec<T> = (0..T::BIT_SIZE)
+            .map(|k| {
+                // we extract a bit of `self.inner()` in position `i` (counting from the left) and
+                // then left-shift that bit by `i`;
+                let mut bits = vec![false; T::BIT_SIZE as usize];
+                bits[k as usize] = self.inner().get_bit_be(k);
+                T::from_bits_be(&bits)
+            })
+            .collect();
+
+        let mut b0 = vec![T::zero(); T::BIT_SIZE as usize];
+        for ((b, c), m) in b0.iter_mut().zip(components.iter()).zip(masks.iter()) {
+            *b = (*c * random) + *m;
+        }
+
+        let mut b1 = vec![T::zero(); T::BIT_SIZE as usize];
+        for (k, ((b, c), m)) in b1
+            .iter_mut()
+            .zip(components.iter())
+            .zip(masks.iter())
+            .enumerate()
+        {
+            *b = ((*c + T::two_pow(T::BIT_SIZE - k as u32 - 1)) * random) + *m;
```

Comment:
It makes it harder to audit the correstness of this algo, since the paper  says
![ddd](https://user-images.githubusercontent.com/555272/217455514-02c357d9-7977-41b5-bc75-5dc4df22d067.png)

With this approach we have the extra burden of explaining why it is ok to make this modification